File: Cpp_src/easy/palindrome_linked_list_234.cc
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// Use iterative:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 20 ms, faster than 93.63%
// Memory Usage: 11.9 MB, less than 100.00%
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        if (head == NULL)
        {
            return true;
        }
        ListNode *slow = head;
        ListNode *fast = head;

        while (fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        slow = reverse(slow);

        while (slow)
        {
            if (slow->val != head->val)
            {
                return false;
            }
            slow = slow->next;
            head = head->next;
        }

        return true;
    }

    ListNode *reverse(ListNode *head)
    {
        ListNode *pre = NULL;
        ListNode *cur = head;

        while (cur)
        {
            ListNode *tmp = cur->next;
            cur->next = pre;
            pre = cur;
            cur = tmp;
        }

        return pre;
    }
};

File: Golang_src/easy/palindromeLinkedList_234.go
package easy

// Use reverse linkedlist & two pointer:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 160 ms, faster than 95.94%
// Memory Usage: 10.9 MB, less than 54.26%
func isPalindromeLinkedList(head *ListNode) bool {
	slow, fast := head, head

	// O(n)
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	// O(n)
	half := revertLinkedList(slow)

	// O(n)
	cur := head
	for half != nil {
		if cur.Val != half.Val {
			return false
		}
		cur = cur.Next
		half = half.Next
	}

	return true
}

func revertLinkedList(head *ListNode) *ListNode {
	var pre *ListNode
	cur := head

	for cur != nil {
		tmp := cur.Next
		cur.Next = pre
		pre = cur
		cur = tmp
	}

	return pre
}
